File: Interval/757. Set Intersection Size At Least Two/solution.py
class Solution:
    def intersectionSizeTwo(self, intervals: List[List[int]]) -> int:
        intervals.sort(key=lambda x:x[1])

        n = len(intervals)

        overlapping_groups = [] # [containing_size, overlap1, overlap2] where overlap1 < overlap2
        for i in range(n):
            if not overlapping_groups or overlapping_groups[-1][2] < intervals[i][0]:
                overlapping_groups.append([2, intervals[i][1]-1, intervals[i][1]])
            else:
                size, overlap1, overlap2 = overlapping_groups[-1]
                start, end = intervals[i]
                if start <= overlap1: continue
                
                if end == overlap2:
                    overlapping_groups[-1] = [size+1, end-1, end]
                else:
                    overlapping_groups[-1] = [size+1, overlap2, end]

        return sum(group[0] for group in overlapping_groups)

File: Interval/757. Set Intersection Size At Least Two/solution.cpp
#include <vector>
using namespace std;


class Solution {
public:
    int intersectionSizeTwo(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end(), [](const auto& lhs, const auto& rhs) {
            if (lhs[1] == rhs[1]) return lhs[0] > rhs[0];
            return lhs[1] < rhs[1];
        });

        vector<int> arr = {intervals[0][1]-1, intervals[0][1]};
        for (int i=1; i<intervals.size(); i++) {
            int l = intervals[i][0], r = intervals[i][1];
            int n = arr.size();
            if (l <= arr[n-2]) continue; // overlap 2
            if (l <= arr[n-1]) { // overlap 1
                arr.push_back(r);
            } else {
                arr.push_back(r-1);
                arr.push_back(r);
            }
        }
        return arr.size();
    }
};

// 但其實我們只在意前個interval, 所以我們可以用O(1) space就好

class Solution {
public:
    int intersectionSizeTwo(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end(), [](const auto& lhs, const auto& rhs) {
            if (lhs[1] == rhs[1]) return lhs[0] > rhs[0];
            return lhs[1] < rhs[1];
        });

        int prevL = intervals[0][1]-1, prevR = intervals[0][1];
        int count = 2;
        for (int i=1; i<intervals.size(); i++) {
            int l = intervals[i][0], r = intervals[i][1];
            
            if (l <= prevL) continue; // overlap 2
            if (l <= prevR) { // overlap 1
                prevL = prevR;
                prevR = r;
                count++;
            } else {
                prevL = r-1;
                prevR = r;
                count += 2;
            }
        }
        return count;
    }
};